packages/react/src/ReactElement.js

```diff
@@ -164,6 +164,91 @@ const ReactElement = function(type, key, ref, self, source, owner, props) {
   return element;
 };
 
+/**
+ * https://github.com/reactjs/rfcs/pull/107
+ */
+export function jsx(type, config, maybeKey /*, source, self */) {
+  let propName;
+
+  // Reserved names are extracted
+  const props = {};
+
+  let key = null;
+  let ref = null;
+  let self = null;
+  let source = null;
+
+  // TODO: I don't allow null/undefined at dev time
+  // I think that decision makes sense because the goal is to pass props
+  // straight through. Is it React's responsibility to create an empty object
+  // if you pass null/undefined?
+  if (config != null) {
+    if (hasValidRef(config)) {
+      ref = config.ref;
+    }
+
+    if (hasValidKey(config)) {
+      key = '' + config.key;
+    }
+
+    // Remaining properties are added to a new props object
+    for (propName in config) {
+      if (
+        hasOwnProperty.call(config, propName) &&
+        !RESERVED_PROPS.hasOwnProperty(propName)
+      ) {
+        props[propName] = config[propName];
+      }
+    }
+  }
+
+  if (key === null && maybeKey !== undefined) {
+    key = maybeKey;
```

Comment:
So there's an interesting case here. In the intermediate step we can't actually support both `<div {...props} key="key" />` and `<div key="key" {...props} />` at the same time here. In case the props has a `key`. They have different semantics but we don't have a way to distinguish them in this function. Therefore I think we should only allow one and have the other compile to createElement and warn in the compiler/lint (or fail to compile).

I believe it makes the most sense that the new function only supports the `<div {...props} key="key" />` form. That way the known static key always win in terms of the React key and it continues to work the same when we update to the new semantics that doesn't allow a key to be spread in.

That means that this block should always override, even if the `key` is non-null here.

Reply:
Great. Makes sense. Should we have the `'' + key` casting behavior too? I think since the key is often a variable the transform will have a hard time verifying that it's a string - so it's not as easy to manage as the non-nullable config/props. I'll add it as well.